### DISCUSSION TITLE
Batch extract dominant colors

### DISCUSSION THREAD

**Discussion by HameedFarah** [ImageMagick/ImageMagick#5100] [Help] [answered]
Hello 
How can I extract the dominant colors from many images into a csv file that contains the name of the file and the main color names (3 at most)

**Comment by fmw42** [comment]
What version of Imagemagick and what platform/OS?  This task will entail scripting to read and write to and from your cdv file. Dominant colors can be extracted via -colors N or better using -kmeans N if on Imagemagick 7.

**Reply by HameedFarah** [reply]
I am on Windows 10
Thank you

**Comment by snibgo** [comment]
I show a method for this, on Windows, at ['Jump' method](http://im.snibgo.com/jumpmeth.htm).

**Reply by HameedFarah** [reply]
Thank you for the link. I took a look at it and it sounds very useful but I am not sure how I can use it to get what I need :)
In other words what is the command I need to run to extract the names of the 3 dominant colors in a folder of images?

**Comment by snibgo** [comment]
Your question is vague, and I don't know what you want.

What are the "dominant colors" of an image? You want colour names, so perhaps you want names like "red", "magenta" and so on. What is the full list of acceptable names? Is "pink" acceptable, or should that be considered "red"?

> ... the names of the 3 dominant colors in a folder of images?

When you have found how to get the "dominant colors" of one image, do you want to apply that to each image in the folder in turn, and then find the most common colours? Or do you first want to make a single image consisting of all the pixels in all the images in the folder, so you then have a single image, so you can find the "dominant colors" of that image?

EDIT: To aid understanding, I suggest you post example images, and state what the dominant colours are. How did you decide that those are the dominant colours?

**Reply by HameedFarah** [reply]
1. Here are the colors options,
Beige
black
blue
brown
green
grey
orange
Pink
Purple
Red
White
Yellow
2. I want to extract a list of each image colors out of hundreds of photos. (apply it to each image in a folder in turn and then find the most colours) for each photo separately.

3. 
This is pink and blue
![content_geordanna-cordero-fields-762612-unsplash](https://user-images.githubusercontent.com/92652480/166841857-4d562bea-4015-42aa-8778-b027afc76a4d.jpg)


This is blue, orange
![images](https://user-images.githubusercontent.com/92652480/166841883-1ecc07df-dbde-4335-a859-265f49678b8f.jpg)

Blue, Red, Orange
![abstract-painting](https://user-images.githubusercontent.com/92652480/166841889-95fdbcf4-d31e-42fb-bb57-9b63f83dad65.jpg)

Blue, Black, Yellow
![FAA_starrry_night](https://user-images.githubusercontent.com/92652480/166841898-76822e1f-7a5a-4e71-8221-476d18e98264.jpg)

**Comment by snibgo** [comment]
Your listed colours are all known to IM; see https://imagemagick.org/script/color.php#color_names .

One approach is to create an image with those colours, and remap the paintings so each pixel becomes either beige, black, blue etc, whichever is the closest.

"Closest" means the smallest Euclidean distance in the colour cube. Doing this in sRGB colorspace is problematic as it isn't perceptually uniform. Lab colorspace is more perceptually uniform.

IM's "blue" is #00f, a very pure blue. The blue in each painting is mostly closer to "purple" than "blue". So I have used "RoyalBlue4" instead.

The code remaps twice: first in sRGB space, then in Lab space. It shows counts of the number of pixels at each remapped colour. It also writes an output image in the remapped colours, so you can judge if it is as you wanted.

I have used "+dither" which turns dithering off. You might prefer it with dithering.

Windows BAT script.
```
%IMG7%magick ^
  xc:Beige xc:black xc:RoyalBlue4 xc:brown xc:green xc:grey ^
  xc:orange xc:Pink xc:Purple xc:Red xc:White xc:Yellow ^
  +append +repage ^
  +write domcolssrgb.png ^
  -colorspace Lab -set colorspace sRGB ^
  +write txt: ^
  domcolslab.png

call :dotest pinkblue.jpg
call :dotest blueorange.jpg
call :dotest blueredorange.jpg
call :dotest blueblackyellow.jpg

exit /B 0

::----------------------------------------------------------
:: Subroutine

:dotest

%IMG7%magick ^
  %1 ^
  +dither -remap domcolssrgb.png ^
  +write %~n1_mapped_srgb.png ^
  -format %%c histogram:info:- |sort

%IMG7%magick ^
  %1 ^
  -colorspace Lab -set colorspace sRGB ^
  +dither -remap domcolslab.png ^
  -set colorspace Lab -colorspace sRGB ^
  +dither -remap domcolssrgb.png ^
  +write %~n1_mapped_lab.png ^
  -format %%c histogram:info:- |sort

exit /B 0
```

**Reply by HameedFarah** [reply]
Thank you so much this is getting much closer now :)
I will need to adjust the color names and add a few more to make it more accurate to what I "see"

Just one last question, how can I generate a text file or CSV file with the file name and the extracted colors?

**Comment by fmw42** [comment]
You can redirect echo or print statements for each file and its best color to a comma delimited text file and then import that into Excel. I expect that snibgo or some other Windows user can show you how.

**Comment by snibgo** [comment]
> Just one last question, how can I generate a text file or CSV file with the file name and the extracted colors?

That is a question about the shell (eg Windows BAT), not ImageMagick.

The obvious way is to parse the text output with "for", and "echo" whatever you want, appending to a text file.
```
for /F "usebackq tokens=1,5,* delims=:(), " %%A in (`%IMG7%magick ^
  %1 ^
  -colorspace Lab -set colorspace sRGB ^
  -remap domcolslab.png ^
  -set colorspace Lab -colorspace sRGB ^
  -remap domcolssrgb.png ^
  +write %~n1_mapped_lab.png ^
  -format %%c histogram:info:- ^|sort`) do (
  echo %~n1,%%A,%%C >>dominants.csv
)
```

**Reply by HameedFarah** [reply]
Thanks again but where exactly do I add this code?
I tried adding it or replacing the routine but it is giving me an error ".jpg was not expected at this time"

**Comment by snibgo** [comment, marked as answer]
Within the ":dotest" subroutine, I showed two alternative "magick" commands. They both write text data about the count of pixels at each colours. The first does the remapping in sRGB space, the second in Lab space.

You asked how to write the data to a CSV file. I modified the second "magick" command to do that. So a complete script might be:
```
del dominants.csv

%IMG7%magick ^
  xc:Beige xc:black xc:RoyalBlue4 xc:brown xc:green xc:grey ^
  xc:orange xc:Pink xc:Purple xc:Red xc:White xc:Yellow ^
  +append +repage ^
  +write domcolssrgb.png ^
  -colorspace Lab -set colorspace sRGB ^
  +write txt: ^
  domcolslab.png

call :dotest pinkblue.jpg
call :dotest blueorange.jpg
call :dotest blueredorange.jpg
call :dotest blueblackyellow.jpg

exit /B 0

::----------------------------------------------------------
:: Subroutine

:dotest

for /F "usebackq tokens=1,5,* delims=:(), " %%A in (`%IMG7%magick ^
  %1 ^
  -colorspace Lab -set colorspace sRGB ^
  -remap domcolslab.png ^
  -set colorspace Lab -colorspace sRGB ^
  -remap domcolssrgb.png ^
  +write %~n1_mapped_lab.png ^
  -format %%c histogram:info:- ^|sort`) do (
  echo %~n1,%%A,%%C >>dominants.csv
)

exit /B 0
```
The output dominants.csv is:
```
pinkblue,24,white 
pinkblue,31,orange 
pinkblue,40,green 
pinkblue,970,black 
pinkblue,1727,srgb(39,64,139) 
pinkblue,13412,purple 
pinkblue,26753,brown 
pinkblue,44758,beige 
pinkblue,87039,pink 
pinkblue,225246,grey 
blueorange,2,yellow 
blueorange,105,red 
blueorange,230,orange 
blueorange,307,green 
blueorange,401,purple 
blueorange,1580,brown 
blueorange,1843,pink 
blueorange,2513,black 
blueorange,3074,beige 
blueorange,4296,white 
blueorange,15718,srgb(39,64,139) 
blueorange,20256,grey 
blueredorange,1222,purple 
blueredorange,2154,pink 
blueredorange,4033,yellow 
blueredorange,8530,beige 
blueredorange,11969,green 
blueredorange,14736,grey 
blueredorange,15107,white 
blueredorange,24365,orange 
blueredorange,27995,red 
blueredorange,111363,srgb(39,64,139) 
blueredorange,128498,brown 
blueredorange,133423,black 
blueblackyellow,3,white 
blueblackyellow,77,pink 
blueblackyellow,318,purple 
blueblackyellow,1437,yellow 
blueblackyellow,2044,beige 
blueblackyellow,2718,brown 
blueblackyellow,3628,orange 
blueblackyellow,18986,green 
blueblackyellow,61779,black 
blueblackyellow,87564,grey 
blueblackyellow,153606,srgb(39,64,139) 
```

**Reply by HameedFarah** [reply]
Thank you so much!
You are amazing
